builder/simplify.cc

```diff
@@ -1040,6 +1040,157 @@ interval_expr bounds_of(const interval_expr& x, const bounds_map& expr_bounds) {
   }
 }
 
+namespace {
+
+class constant_bound : public node_mutator {
+  // 1 -> we are looking for an upper bound
+  // 0 -> we are looking for a lower bound
+  int sign;
+
+public:
+  constant_bound(int sign) : sign(sign) {}
+
+  void visit(const class min* op) override {
+    if (sign < 0) {
+      // We can only learn about upper bounds from min.
+      set_result(op);
+      return;
+    }
+    expr a = mutate(op->a);
+    expr b = mutate(op->b);
+    const index_t* ca = as_constant(a);
+    const index_t* cb = as_constant(b);
+    if (ca && cb) {
+      set_result(expr(std::min(*ca, *cb)));
+    } else if (ca) {
```

Comment:
I don't follow the logic here -- if we have `min(constant, expr)`, we resolve to just `constant`? Either this is wrong or I misunderstand the purpose of this mutator.

Reply:
The goal of this mutator is to find constant upper bounds. So if we see something like `min(x, 3)`, we know the expression has an upper bound of 3.